clash-prelude/src/Clash/Promoted/Nat.hs

```diff
@@ -404,14 +437,28 @@ showBNat = go []
 -- | Convert a singleton natural number to its base-2 representation
 --
 -- __NB__: Not synthesizable
-toBNat :: SNat n -> BNat n
+toBNat :: forall n. SNat n -> BNat n
+#if MIN_VERSION_base(4,16,0)
+toBNat s@SNat = case cmpNat (SNat @1) s of
+  LTI -> case euclideanNat @2 @n of
+    Sub Dict -> case sameNat (SNat @(n `Mod` 2)) (SNat @0) of
+      Just Refl -> B0 (toBNat (SNat @(n `Div` 2)))
+      Nothing -> case sameNat (SNat @(n `Mod` 2)) (SNat @1) of
+        Just Refl -> B1 (toBNat (SNat @(n `Div` 2)))
```

Comment:
Maybe you could share the `mod 2` and `div 2` binders between these assignments to halve the time the plugin will spend on these computations? (I'm assuming this isn't done automagically, which is an assumption without a lot of clout behind it, I'll admit it)